### ISSUE TITLE
apache and nginx plugin require snakeoil cert and key

### ISSUE THREAD

**Issue by kuba** [letsencrypt/letsencrypt#481]
1. Nginx plugin [requires `/etc/ssl/certs/ssl-cert-snakeoil.pem` and `/etc/ssl/private/ssl-cert-snakeoil.key`](https://github.com/letsencrypt/lets-encrypt-preview/blob/9394c26a9b532b6b8b5df1735114d60e5a1a9e27/letsencrypt_nginx/configurator.py#L273-278).
2. Nginx plugin fails whenever `nginx` fails to restart.
3. Nginx fails to restart when config file contains broken paths.
4. Snakeoil cert and key are created in post install hook of the `ssl-cert` Debian package. This package is not installed by `./bootstrap/debian.sh`.
5. Admin might simply delete snakeoil cert and key anyway.
6. We want to support systems other than Debian-derviatives.

Due to 5. and 6. it's not sufficient to update bootstrap scripts to `apt-get install ssl-cert`. I would suggest creating our own snakeoil certs in the process or deleting those `ssl_certificate` and `ssl_certificate_key` stanzas altogether. @diracdeltas?

Tested on fresh digitalocean 14.04 x64 droplet, reported by @jcjones.

**Comment by diracdeltas** [comment]
Good find. The snakeoil certs requirement was copied from what Apache did (or what I thought it did) at the time. I have a feeling that deleting the `ssl_certificate` and `ssl_certificate_key` directives would cause Nginx to fail to restart; maybe that's okay.

Creating our own snakeoil certs on the fly seems fine.

**Comment by kuba** [comment]
Apache is also affected.

**Comment by jdkasten** [comment]
The Apache vhost isn't enabled until it has a valid cert. So I don't
believe it is affected.
Though Apache is a work in progress and will be under lots of review.

**Comment by jdkasten** [comment]
Confirmed that Apache is not affected.

**Comment by kuba** [comment]
@jdkasten, apache plugin explicitly refers to `/etc/ssl/certs/ssl-cert-snakeoil.pem` and `/etc/ssl/private/ssl-cert-snakeoil.key` at https://github.com/letsencrypt/letsencrypt/blob/f99af51de314b66ab9ab7e4d8a88205a00a74e40/letsencrypt_apache/configurator.py#L530-534. Dead code?

**Comment by jdkasten** [comment]
No... that is code inside make_vhost_ssl.  Which is called when an ssl vhost from sites-available (not enabled) can't be located.

This creates a dummy ssl vhost file within "sites-available", but it does not enable it.  It does not put it into the configuration and the webserver is not affected.

Before it is enabled/deployed/used for anything, the cert and key directives are replaced.
Here https://github.com/letsencrypt/letsencrypt/blob/master/letsencrypt_apache/configurator.py#L159

Once the directives have been changed, the vhost is deployed (enabled) here.
https://github.com/letsencrypt/letsencrypt/blob/master/letsencrypt_apache/configurator.py#L217

Apache configurations often start with a "default-ssl" available host which contains the snake-oil parameters. This isn't a problem because the site isn't enabled or loaded into the web server.

The code that is currently there applies the same logic.

**Comment by jdkasten** [comment]
The code will get cleaned up though.  I haven't had time to touch the logic. It hasn't been updated in years.

**Comment by kuba** [comment]
Okay, so it seems the fix would be to use:

```
        # certificate and key will be updated with real paths in ...
        self.parser.add_dir(vh_p[0], "SSLCertificateFile", "not used yet")
        self.parser.add_dir(vh_p[0], "SSLCertificateKeyFile", "not used yet")
        self.parser.add_dir(vh_p[0], "Include", self.parser.loc["ssl_options"])
```

What do you think?

**Comment by kuba** [comment]
@jdkasten, shall we reopen this to track progress on Apache and the removal of slightly confusing statements?

**Comment by jdkasten** [comment]
I agree that the directives should be changed, but the title of this bug does not reflect the state of the code.  This actually follows Debian practice in creating an example ssl vhost file.  The snakeoil cert/key are not required, it will never appear in the installed configuration.

There are much larger problems that need to be addressed with Apache that I am currently tracking down.  I did make a new issue for this though, just so it doesn't get lost.  (I could file about 10 that I am working on.) https://github.com/letsencrypt/letsencrypt/issues/571

I changed the name to include Alpha... as I know the code will not work in all cases and the code needs some restructuring.  I am working on it.

https://github.com/letsencrypt/letsencrypt/blob/master/letsencrypt_apache/configurator.py#L89
